### PR TITLE
Fix __all__ issues

### PR DESCRIPTION
**Pull request by JelleZijlstra**

### CODE REVIEW

**Review by AlexWaygood — COMMENTED** [review]
Looks like this fixes the issue, but also creates a new issue where `typeshed_client` chokes on typeshed's latest `asyncio` stubs:

```pycon
>>> from pathlib import Path
>>> import typeshed_client
>>> context = typeshed_client.finder.get_search_context(typeshed=Path("../typeshed/stdlib"), search_path=[], version=(3, 10))
>>> typeshed_client.get_stub_names("asyncio", search_context=context)
could not import ('asyncio', 'base_events') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'coroutines') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'events') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'futures') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'locks') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'protocols') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'queues') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'runners') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'streams') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'subprocess') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'tasks') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'transports') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'exceptions') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'threads') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
could not import ('asyncio', 'windows_events') in ('asyncio',) with SearchContext(typeshed=WindowsPath('../typeshed/stdlib'), search_path=[], version=(3, 10), platform='win32')
{'sys': NameInfo(name='sys', is_exported=False, ast=ImportedName(module_name=('sys',), name=None), child_nodes=None)}
```

File: tests/test.py [hunk 1]
```diff
@@ -150,6 +150,26 @@ def test_starimport(self) -> None:
...
+        for name in expected:
+            self.check_nameinfo(names, name, typeshed_client.ImportedName)
+            path = typeshed_client.ModulePath(("dunder_all",))
+            self.assertEqual(names[name].ast, typeshed_client.ImportedName(path, name))
+
+    def test_starimport_no_dunders(self) -> None:
+        ctx = get_context((3, 10))
+        names = get_stub_names("importabout", search_context=ctx)
+        assert names is not None
+        self.assertEqual(set(names.keys()), {"x"})
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply tests/test.py:168]: Fixed the other 10 instances of this too :)

File: typeshed_client/parser.py [hunk 2]
```diff
@@ -115,6 +115,46 @@ def parse_ast(
...
+    if isinstance(info.ast, ImportedName):
+        raise InvalidStub(f"Invalid __all__: {info}")
+    return _get_dunder_all_from_ast(info.ast)
+
+
+def _get_dunder_all_from_ast(node: ast.AST) -> Optional[List[str]]:
+    if not isinstance(node, (ast.Assign, ast.AugAssign)):
+        return None
+    rhs = node.value
+    if not isinstance(rhs, ast.List):
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply typeshed_client/parser.py]: Good point, added a test for this too.

File: typeshed_client/parser.py [hunk 3]
```diff
@@ -115,6 +115,46 @@ def parse_ast(
...
+            names += subnames
+        return names
+    if isinstance(info.ast, ImportedName):
+        raise InvalidStub(f"Invalid __all__: {info}")
+    return _get_dunder_all_from_ast(info.ast)
+
+
+def _get_dunder_all_from_ast(node: ast.AST) -> Optional[List[str]]:
+    if not isinstance(node, (ast.Assign, ast.AugAssign)):
+        return None
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply typeshed_client/parser.py]: Changed.

**Comment by JelleZijlstra** [comment]
Thanks, that was very useful feedback!

I think the tuple change also fixes what you saw with asyncio, I can't reproduce that any more.

I'll look into making the integration test fail if there are any warnings.

**Comment by AlexWaygood** [comment]
> Thanks, that was very useful feedback!

No worries -- thanks for the quick fix!

**Comment by AlexWaygood** [comment]
> I think the tuple change also fixes what you saw with asyncio, I can't reproduce that any more.

Yup, confirmed that this is now fixed on `master`!

**Pull request merged**